src/argo.hpp

```diff
@@ -69,6 +70,41 @@ namespace argo {
 	 */
 	int number_of_nodes();
 
+	/**
+	 * @brief Check if addr belongs in the ArgoDSM memory space
+	 * @tparam T type of addr
+	 * @param addr A memory address
+	 * @return True if addr is in the ArgoDSM memory space, else false
+	 */
+	template<typename T>
+	bool is_argo_address(T* addr) {
+		char* check_addr = static_cast<char*>(addr);
+		char* argo_start = backend::global_base();
+		char* argo_end = argo_start + backend::global_size();
+		return static_cast<bool>(check_addr >= argo_start && check_addr < argo_end);
+	}
+
+	/**
+	 * @brief Get the ArgoDSM node id where addr is physically backed
+	 * @param addr A valid address in the ArgoDSM memory space
+	 * @tparam T type of addr
+	 * @return The id of the node physically backing addr, or
+	 * argo::data_distribution::invalid_node_id if it has not yet
+	 * been first-touched under the first-touch allocation policy
+	 * @pre addr must be an address in ArgoDSM memory
+	 */
+	template<typename T>
+	int get_homenode(T* addr) {
+		data_distribution::global_ptr<T> gptr(addr, false, false);
+		return data_distribution::is_first_touch_policy() ?
+			gptr.peek_node() : gptr.node();
```

Comment:
should `peek_node` not always be "correct" here? my understanding is that `node` causes a homenode to be forced to exist while `peek_node` can return an invalid one iff it is not yet set. As for non-first-touch there is always a homenode set, `peek_node` should return the same value?

Reply:
I will rewrite the `peek_homenode` implementation of each policy that directly calculates homenodes to call `homenode`, and alter this accordingly.

Reply:
Should be fixed now.